test/test_client.rb

```diff
@@ -122,6 +122,83 @@ class QueuedWorker
       assert QueuedWorker.perform_async(1, 2)
       assert_equal 1, Sidekiq::Queue.new('flimflam').size
     end
+
+    describe 'argument checking' do
+      class InterestingWorker
+        include Sidekiq::Worker
+
+        def perform(a_symbol, a_date, a_hash, a_struct)
+        end
+      end
+
+      class InterestingDeepWorker
+        include Sidekiq::Worker
+
+        def perform(a_deep_structure)
+        end
+      end
+
+      it 'enqueues jobs with interesting arguments' do
+        InterestingWorker.perform_async(
+          :symbol,
+          Date.new(2021, 1, 1),
+          { some: 'hash', 'with' => 'different_keys' },
+          Struct.new(:x, :y).new(0, 0)
+        )
+      end
+
+      it 'works with a JSON-friendly deep, nested structure' do
+        InterestingDeepWorker.perform_async(
+          {
+            'foo' => ['a', 'b', 'c'],
+            'bar' => ['x', 'y', 'z']
+          }
+        )
+      end
+
+      describe 'strict mode is enabled' do
+        before do
+          Sidekiq.strict_mode!
+        end
+
+        after do
+          Sidekiq.strict_mode!(false)
+        end
+
+        it 'raises an error' do
+          assert_raises ArgumentError do
+            InterestingWorker.perform_async(
+              :symbol,
+              Date.new(2021, 1, 1),
+              { some: 'hash', 'with' => 'different_keys' },
```

Comment:
I'd prefer to see a few different tests to verify we catch symbols in Hashes, symbols in Arrays and Dates. As is, this test fails but maybe because it catches one of those cases but not the others. Test one simple case at a time.